src/bidiMapper/domains/context/BrowsingContextImpl.ts

```diff
@@ -704,52 +707,105 @@ export class BrowsingContextImpl {
     await this.#cdpTarget.cdpClient.sendCommand('Page.bringToFront');
   }
 
-  async captureScreenshot(): Promise<BrowsingContext.CaptureScreenshotResult> {
+  async captureScreenshot(
+    params: BrowsingContext.CaptureScreenshotParameters
+  ): Promise<BrowsingContext.CaptureScreenshotResult> {
+    if (!this.isTopLevelContext()) {
+      throw new UnsupportedOperationException(
+        `Non-top-level 'context' is currently not supported`
+      );
+    }
+
     // XXX: Focus the original tab after the screenshot is taken.
     // This is needed because the screenshot gets blocked until the active tab gets focus.
     await this.#cdpTarget.cdpClient.sendCommand('Page.bringToFront');
 
-    let clip: Protocol.DOM.Rect;
-
-    if (this.isTopLevelContext()) {
-      const {cssContentSize, cssLayoutViewport} =
-        await this.#cdpTarget.cdpClient.sendCommand('Page.getLayoutMetrics');
-      clip = {
-        x: cssContentSize.x,
-        y: cssContentSize.y,
-        width: cssLayoutViewport.clientWidth,
-        height: cssLayoutViewport.clientHeight,
-      };
-    } else {
-      const {
-        result: {value: iframeDocRect},
-      } = await this.#cdpTarget.cdpClient.sendCommand(
-        'Runtime.callFunctionOn',
-        {
-          functionDeclaration: String(() => {
-            const docRect =
-              globalThis.document.documentElement.getBoundingClientRect();
-            return JSON.stringify({
-              x: docRect.x,
-              y: docRect.y,
-              width: docRect.width,
-              height: docRect.height,
-            });
-          }),
-          executionContextId: this.#defaultRealm.executionContextId,
+    const sandbox = await this.getOrCreateSandbox(undefined);
+    let clip = await (async () => {
+      const {clip} = params;
+      if (!clip) {
+        return;
+      }
+      switch (clip.type) {
+        case 'viewport':
+          return {x: clip.x, y: clip.y, width: clip.width, height: clip.height};
+        case 'element': {
+          const result = await sandbox.callFunction(
+            String((element: unknown) => {
+              return element instanceof Element;
+            }),
+            {type: 'undefined'},
+            [clip.element],
+            false,
+            Script.ResultOwnership.None,
+            {}
+          );
+          if (result.type === 'exception') {
+            throw new NoSuchElementException(
+              `Element ${clip.element.sharedId} was not found`
+            );
+          }
+          assert(result.result.type === 'boolean');
+          if (!result.result.value) {
+            throw new NoSuchElementException(
+              `Node ${clip.element.sharedId} is not an Element`
+            );
+          }
+          if (clip.scrollIntoView) {
+            throw new UnsupportedOperationException(
```

Comment:
Extract this before the switch statement.
Also add this to https://github.com/GoogleChromeLabs/chromium-bidi/issues/514 if you are not gonna work on it next. Same for Top-level Context

Reply:
Cannot because we type discriminate here.